### ISSUE TITLE
automatically select matching borg version

### ISSUE THREAD

**Issue by leo-b** [borgbackup/borg#1392]
In a heterogeneous environment with multiple clients, it can be useful to have multiple borg versions installed on the backup server and let the client choose the matching remote version.

This could be obtained by introducing a new placeholder {borgversion} and add placeholder substitution to --remote-path and the BORG_RSH environment. Another way is to pass the version number via an environment variable:

- Adding a version selector like --remote-path "/usr/bin/borg-{borgversion}" is useful for direct selection of the server binary.

- If the remote command is locked to a fixed command server-side by "command=..." in _.ssh/authorized_keys_, the BORG_VERSION environment-variable may be passed:
  - client side: add ``SendEnv BORG_VERSION`` to _.ssh/config_ or use ``BORG_RSH=ssh -oSendEnv=BORG_VERSION``.
  - server side:
    - use ``AcceptEnv BORG_VERSION`` in _sshd_config_.
    - use a _.ssh/authorized_keys_ file like that: ``command="/home/borg/bin/borgserver.sh --restrict-to-path ...",no-pty,no-agent-forwarding,no-port-forwarding,no-X11-forwarding,no-user-rc ssh-rsa ...``
    - _/home/borg/bin/borgserver.sh_:
      ```
      #!/bin/bash

      BORGPROG="borg"
      if [ -n "$BORG_VERSION" ]; then
        if [[ $BORG_VERSION =~ ^[0-9]+[._0-9]*$ ]]; then
          BORGPROG="borg-$BORG_VERSION"
        else
          exit 1
        fi
      fi

      BORGPATH="/home/borg/bin/$BORGPROG"
      exec "$BORGPATH" serve "$@"
      ```

**Comment by ThomasWaldmann** [comment]
So, why do you process BORG_RSH with placeholder expansion?

**Comment by PlasmaPower** [comment]
> use AcceptEnv BORG_VERSION in sshd_config

I'd prefer a solution without environment changes so that it doesn't require this. While users can modify their own authorized_keys, they can't modify their part of sshd_config (on a normal setup at least).

**Comment by leo-b** [comment]
> So, why do you process BORG_RSH with placeholder expansion?

It is now optional. I did it before I added the BORG_VERSION environment variable. It allows setups like
```
BORG_RSH="env BORG_VERSION={borgversion} ssh -oSendEnv=BORG_VERSION"
```
... which is redundant now, if BORG_VERSION is always defined in ssh's environment.

**Comment by leo-b** [comment]
> I'd prefer a solution without environment changes.

Then choose my first suggested solution. (``--remote-path "/usr/bin/borg-{borgversion}"``)

But as this is a valid use case, there is also the case where backup admins don't like the user to be able to specify a remote ssh command and thus lock it as outlined in my second use case. So I believe both scenarios have their respective merits.

**Comment by ThomasWaldmann** [comment]
Please give a now-valid use case of expanding BORG_RSH (with borgversion or whatever).

**Comment by leo-b** [comment]
> Please give a now-valid use case of expanding BORG_RSH

With BORG_VERSION being set and the current supported placeholders, I can't think of any good use case anymore.
So I can remove the placeholder substitution of BORG_RSH if you don't like it...

**Comment by ThomasWaldmann** [comment]
Well, if we can't think of any use case, we maybe don't need it. Maybe wait 1-2d for feedback of others.

**Comment by ThomasWaldmann** [comment]
Fixed by #1423.
